--- ./CcspWifiAgent/include/TR-181/ml/ccsp_WifiLog_wrapper.h
+++ ./CcspWifiAgent/include/TR-181/ml/ccsp_WifiLog_wrapper.h	2015-04-22 14:22:31.190835008 -0600
@@ -0,0 +1,27 @@
+#ifndef  _CCSP_WIFILOG_WRPPER_H_ 
+#define  _CCSP_WIFILOG_WRPPER_H_
+
+#include "ccsp_custom_logs.h"
+extern ANSC_HANDLE bus_handle;
+extern char g_Subsystem[32];
+/*
+ * Logging wrapper APIs g_Subsystem
+ */
+#define  CcspTraceBaseStr(arg ...)                                                             \
+            do {                                                                            \
+                snprintf(pTempChar1, 4095, arg);                                         	\
+            } while (FALSE)
+			
+#define  CcspWifiTrace(msg)                         \
+{\
+	                char*   pTempChar1 = (char*)malloc(4096);                                     \
+                if ( pTempChar1 )                                                             \
+                {                                                                            \
+					char	sfn[32];														 \
+                    CcspTraceBaseStr msg;                                                       \
+				    WriteLog(pTempChar1,bus_handle,g_Subsystem,"Device.LogAgent.WifiLogMsg"); \
+                    free(pTempChar1);                                                         \
+                }\
+}				
+ //   WriteLog(msg,bus_handle,g_Subsystem,"Device.LogAgent.WifiLogMsg")
+#endif
\ No newline at end of file
